### PR TITLE
feat(dev): Add support for pyenv v2 and drop support for pyenv v1

### PR DESCRIPTION
**Pull request by armenzg**
The latest release of pyenv does not allow using `pyenv init -`, instead,
it requires using `--path` (i.e. `pyenv init --path`).

Anyone calling `make setup-pyenv` will have to update to the latest `pyenv`.

Not fixing this currently affects new engineers when they set up their computers.

### CODE REVIEW

**Comment by armenzg** [comment]
I've decided to separate PR #26296 and this one.
They should land soon after the other.

File: scripts/pyenv_setup.sh [hunk 1]
```diff
@@ -60,12 +62,14 @@ _append_to_startup_script() {
...
   if [[ -n "$1" ]]; then
-    echo "Adding pyenv init (if missing) to ${1}..."
-    # shellcheck disable=SC2016
-    if ! grep -qF "pyenv init" "${1}"; then
-      # pyenv init - is needed to include the pyenv shims in your PATH
-      # The first \n is very important since on Github workers the output was being appended to
-      # the last line rather than on a new line. I never figured out why
+    if grep -qF "(pyenv init -)" "${1}"; then
+      echo >&2 "!!! Please remove the old-style pyenv initialization and try again:"
+      echo "sed -i.bak 's/(pyenv init -)/(pyenv init --path)/' ${1}"
```
Comment (armenzg) [inline scripts/pyenv_setup.sh:67]: Should we just do this automatically and not interrupt the user?

File: .github/workflows/development-environment.yml [hunk 2]
```diff
@@ -46,14 +46,15 @@ jobs:
...
         id: python-version
-        with:
-          python-version: 3.6.13
+        # XXX: Can we cache the installed Python environments?
+        # We echo to $GITHUB_ENV to make the PATH changes by pyenv init permanent for the rest of the execution
+        run: |
+          make setup-pyenv
+          eval "$(pyenv init --path)"
+          echo "PATH=$PATH" >> $GITHUB_ENV
+          echo "::set-output name=python-version::$(python -V  | sed s/Python\ //g)"
```
Comment (armenzg) [inline .github/workflows/development-environment.yml:57]: This block is simply to undo what we temporarily add on PR #26296.
Now that pyenv is fixed we can now call `make setup-pyenv` again.

**Comment by billyvg** [comment]
Is there a changelog for pyenv2? Should we move to v2 or keep people on v1?

**Review by joshuarli — APPROVED** [review]

**Comment by armenzg** [comment]
> Is there a changelog for pyenv2? Should we move to v2 or keep people on v1?

@billyvg the release info:
https://github.com/pyenv/pyenv/releases/tag/v2.0.0

If people are calling `make setup-pyenv` I want them to upgrade so we don't have to support multiple set ups.

The one that matters for M1 is:
> Support for CPython 3.8.10 has been added.

The only known breaking change is (which is what we handle in this PR IIUC):
>  Split startup logic into PATH and everything else (#1649 (comment))

**Review by billyvg — APPROVED** [review]

**Pull request merged**